src/librustc_resolve/build_reduced_graph.rs

```diff
@@ -639,7 +639,10 @@ impl<'a> Resolver<'a> {
         // but metadata cannot encode gensyms currently, so we create it here.
         // This is only a guess, two equivalent idents may incorrectly get different gensyms here.
         let ident = ident.gensym_if_underscore();
-        let def_id = def.def_id();
+        let def_id = match def.opt_def_id() {
+            Some(def_id) => def_id,
+            None => return,
```

Comment:
Ignoring things without a def-id is not a correct behavior.

`def_id`s should rather be obtained though destructuring inside `match def { ... }`, and new `Def` variants need to be added the match as needed, and they should go through `self.define(...)` too.